aptos-move/aptos-vm/src/data_cache.rs

```diff
@@ -118,15 +118,49 @@ impl<'e, E: ExecutorView> StorageAdapter<'e, E> {
             };
 
             let buf_size = resource_size(&buf);
-            Ok((buf, buf_size + group_size as usize))
-        } else {
-            let state_key = resource_state_key(address, struct_tag)?;
-            let buf = self
-                .executor_view
-                .get_resource_bytes(&state_key, maybe_layout)?;
-            let buf_size = resource_size(&buf);
-            Ok((buf, buf_size))
+            return Ok(Some((buf, buf_size + group_size as usize)));
         }
+        Ok(None)
+    }
+
+    fn get_any_resource_with_layout(
+        &self,
+        address: &AccountAddress,
+        struct_tag: &StructTag,
+        metadata: &[Metadata],
+        maybe_layout: Option<&MoveTypeLayout>,
+    ) -> PartialVMResult<(Option<Bytes>, usize)> {
+        let resource_group_bytes =
+            self.get_resource_group_bytes(address, struct_tag, metadata, maybe_layout)?;
+        if let Some(result) = resource_group_bytes {
+            return Ok(result);
+        }
+
+        let state_key = resource_state_key(address, struct_tag)?;
+        let buf = self
+            .executor_view
+            .get_resource_bytes(&state_key, maybe_layout)?;
+        let buf_size = resource_size(&buf);
+        Ok((buf, buf_size))
+    }
+
+    fn get_any_resource_size_with_layout(
+        &self,
+        address: &AccountAddress,
+        struct_tag: &StructTag,
+        metadata: &[Metadata],
+        maybe_layout: Option<&MoveTypeLayout>,
+    ) -> PartialVMResult<(Option<u64>, usize)> {
```

Comment:
let's define a type (probably in third-party Move where resolver lives) so that it is clear what is what:
```rust
struct {
  // Size of the resource or resource group member. None if it does not exist.
  size: Optionu<u64>
  // Number of bytes loaded when accessing this resource, used for gas charging only.
  bytes_loaded: u64,
}
```
or similar?